messaging/messaging.go

```diff
@@ -0,0 +1,201 @@
+	Data                  map[string]interface{} `json:"data,omitempty"`
+	Notification          AndroidNotification    `json:"notification,omitempty"`
+}
+
+// AndroidNotification is notification to send to android devices.
+// See https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages#AndroidNotification
+type AndroidNotification struct {
+	Title        string   `json:"title,omitempty"`
+	Body         string   `json:"body,omitempty"`
+	Icon         string   `json:"icon,omitempty"`
+	Color        string   `json:"color,omitempty"`
+	Sound        string   `json:"sound,omitempty"`
+	Tag          string   `json:"tag,omitempty"`
+	ClickAction  string   `json:"click_action,omitempty"`
+	BodyLocKey   string   `json:"body_loc_key,omitempty"`
+	BodyLocArgs  []string `json:"body_loc_args,omitempty"`
+	TitleLocKey  string   `json:"title_loc_key,omitempty"`
+	TitleLocArgs []string `json:"title_loc_args,omitempty"`
+}
+
+// WebpushConfig is Webpush protocol options.
+// See https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages#WebpushConfig
+type WebpushConfig struct {
+	Headers      map[string]interface{} `json:"headers,omitempty"`
+	Data         map[string]interface{} `json:"data,omitempty"`
+	Notification WebpushNotification    `json:"notification,omitempty"`
+}
+
+// WebpushNotification is Web notification to send via webpush protocol.
+// See https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages#WebpushNotification
+type WebpushNotification struct {
+	Title string `json:"title,omitempty"`
+	Body  string `json:"body,omitempty"`
+	Icon  string `json:"icon,omitempty"`
+}
+
+// ApnsConfig is Apple Push Notification Service specific options.
+// See https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages#ApnsConfig
+type ApnsConfig struct {
+	Headers map[string]interface{} `json:"headers,omitempty"`
+	Payload map[string]interface{} `json:"payload,omitempty"`
+}
+
+// NewClient creates a new instance of the Firebase Cloud Messaging Client.
+//
+// This function can only be invoked from within the SDK. Client applications should access the
+// the Messaging service through firebase.App.
+func NewClient(ctx context.Context, c *internal.MessagingConfig) (*Client, error) {
+	if c.ProjectID == "" {
+		return nil, errors.New("project id is required to access firebase cloud messaging client")
+	}
+
+	hc, _, err := transport.NewHTTPClient(ctx, c.Opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		endpoint: messagingEndpoint,
+		client:   &internal.HTTPClient{Client: hc},
+		project:  c.ProjectID,
+		version:  "Go/Admin/" + c.Version,
+	}, nil
+}
+
+// SendMessage sends a Message to Firebase Cloud Messaging.
+//
+// Send a message to specified target (a registration token, topic or condition).
+// https://firebase.google.com/docs/cloud-messaging/send-message
+func (c *Client) SendMessage(ctx context.Context, payload *RequestMessage) (msg *ResponseMessage, err error) {
```

Comment:
Just spitballing some ideas here. How about the following API signatures?

```
func (c *Client) Send(ctx context.Context, message *Message) (string, error)
func (c *Client) SendDryRun(ctx context.Context, message *Message) (string, error)
```

This is what I was going for in Node.js and Java environments. We accept a `Message` instead of `RequestMessage`, and do the necessary wrapping in the SDK. Also we return the string ID returned by the FCM server, instead of the parsed JSON struct.

Reply:
OK it sounds good to me, i'm working on it